### ISSUE TITLE
script

### ISSUE THREAD

**Issue by parag-shendye** [parag-shendye/WalnutChat#1]
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# Set random seed for reproducibility
torch.manual_seed(42)
np.random.seed(42)

# Define the Autoencoder model
class TimeSeriesAutoencoder(nn.Module):
    def __init__(self, input_dim, encoding_dim=16):
        super(TimeSeriesAutoencoder, self).__init__()
        
        # Encoder
        self.encoder = nn.Sequential(
            nn.Linear(input_dim, 64),
            nn.ReLU(),
            nn.Linear(64, 32),
            nn.ReLU(),
            nn.Linear(32, encoding_dim),
            nn.ReLU()
        )
        
        # Decoder
        self.decoder = nn.Sequential(
            nn.Linear(encoding_dim, 32),
            nn.ReLU(),
            nn.Linear(32, 64),
            nn.ReLU(),
            nn.Linear(64, input_dim)
        )
    
    def forward(self, x):
        encoded = self.encoder(x)
        decoded = self.decoder(encoded)
        return decoded

# Function to load and preprocess data
def load_and_preprocess_data(file_path, sequence_length=10):
    """
    Load time series data and prepare sequences for autoencoder training
    
    Args:
        file_path: Path to CSV file with columns [distance, sen1, sen2]
        sequence_length: Number of time steps to use in each sequence
        
    Returns:
        X_train, X_test: Training and test data for autoencoder
        scaler: Fitted scaler for inverse transformation
    """
    # Load data
    df = pd.read_csv(file_path)
    
    # Extract features
    features = df[['distance', 'sen1', 'sen2']].values
    
    # Normalize data
    scaler = MinMaxScaler()
    features_scaled = scaler.fit_transform(features)
    
    # Create sequences
    sequences = []
    for i in range(len(features_scaled) - sequence_length + 1):
        sequences.append(features_scaled[i:i+sequence_length].flatten())
    
    sequences = np.array(sequences)
    
    # Split into train and test sets
    X_train, X_test = train_test_split(sequences, test_size=0.2, random_state=42)
    
    return X_train, X_test, scaler, sequence_length

# Function to train the autoencoder
def train_autoencoder(model, X_train, batch_size=32, epochs=100, learning_rate=0.001):
    """
    Train the autoencoder model
    
    Args:
        model: The autoencoder model
        X_train: Training data
        batch_size: Batch size for training
        epochs: Number of training epochs
        learning_rate: Learning rate for optimizer
        
    Returns:
        model: Trained model
        history: Training history (loss values)
    """
    # Convert to PyTorch tensors
    train_tensor = torch.FloatTensor(X_train)
    train_dataset = TensorDataset(train_tensor, train_tensor)
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    
    # Define loss function and optimizer
    criterion = nn.MSELoss()
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
    
    # Training loop
    history = []
    for epoch in range(epochs):
        model.train()
        train_loss = 0.0
        
        for batch_features, _ in train_loader:
            # Forward pass
            outputs = model(batch_features)
            loss = criterion(outputs, batch_features)
            
            # Backward pass and optimize
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            
            train_loss += loss.item()
        
        # Calculate average loss
        train_loss /= len(train_loader)
        history.append(train_loss)
        
        if (epoch+1) % 10 == 0:
            print(f'Epoch {epoch+1}/{epochs}, Loss: {train_loss:.6f}')
    
    return model, history

# Function to detect anomalies
def detect_anomalies(model, X_test, threshold_multiplier=3.0):
    """
    Detect anomalies in test data using reconstruction error
    
    Args:
        model: Trained autoencoder model
        X_test: Test data
        threshold_multiplier: Multiplier for standard deviation to set anomaly threshold
        
    Returns:
        is_anomaly: Boolean array indicating anomalies
        reconstruction_errors: Array of reconstruction errors
        threshold: The threshold value used for anomaly detection
    """
    model.eval()
    test_tensor = torch.FloatTensor(X_test)
    
    # Get reconstructions
    with torch.no_grad():
        reconstructions = model(test_tensor).numpy()
    
    # Calculate reconstruction error
    reconstruction_errors = np.mean(np.square(reconstructions - X_test), axis=1)
    
    # Set threshold as mean + n*std
    threshold = np.mean(reconstruction_errors) + threshold_multiplier * np.std(reconstruction_errors)
    
    # Detect anomalies
    is_anomaly = reconstruction_errors > threshold
    
    return is_anomaly, reconstruction_errors, threshold

# Main function to run the complete pipeline
def main(file_path, sequence_length=10, encoding_dim=16, epochs=100, batch_size=32, threshold_multiplier=3.0):
    """
    Run the complete anomaly detection pipeline
    
    Args:
        file_path: Path to the time series CSV file
        sequence_length: Length of each input sequence
        encoding_dim: Dimension of the encoded representation
        epochs: Number of training epochs
        batch_size: Batch size for training
        threshold_multiplier: Multiplier for anomaly threshold
    """
    print("Loading and preprocessing data...")
    X_train, X_test, scaler, seq_length = load_and_preprocess_data(file_path, sequence_length)
    
    # Initialize model
    input_dim = X_train.shape[1]  # Flattened sequence dimensions
    model = TimeSeriesAutoencoder(input_dim, encoding_dim)
    
    print("Training autoencoder...")
    model, history = train_autoencoder(model, X_train, batch_size, epochs)
    
    print("Detecting anomalies...")
    is_anomaly, errors, threshold = detect_anomalies(model, X_test, threshold_multiplier)
    
    # Print results
    anomaly_count = np.sum(is_anomaly)
    print(f"Total test samples: {len(X_test)}")
    print(f"Anomalies detected: {anomaly_count} ({anomaly_count/len(X_test)*100:.2f}%)")
    print(f"Anomaly threshold: {threshold:.6f}")
    
    # Plot training history
    plt.figure(figsize=(10, 4))
    plt.subplot(1, 2, 1)
    plt.plot(history)
    plt.title('Training Loss')
    plt.xlabel('Epoch')
    plt.ylabel('MSE Loss')
    
    # Plot reconstruction errors
    plt.subplot(1, 2, 2)
    plt.hist(errors, bins=50)
    plt.axvline(x=threshold, color='r', linestyle='--', label=f'Threshold: {threshold:.6f}')
    plt.title('Reconstruction Error Distribution')
    plt.xlabel('Reconstruction Error')
    plt.ylabel('Count')
    plt.legend()
    
    plt.tight_layout()
    plt.savefig('anomaly_detection_results.png')
    plt.show()
    
    return model, is_anomaly, errors, threshold

# Example usage
if __name__ == "__main__":
    # Replace with your file path
    file_path = "your_time_series_data.csv"
    
    # Run the anomaly detection pipeline
    model, is_anomaly, errors, threshold = main(
        file_path=file_path,
        sequence_length=10,
        encoding_dim=16,
        epochs=100,
        batch_size=32,
        threshold_multiplier=3.0
    )
    
    # Save model
    torch.save(model.state_dict(), 'time_series_autoencoder.pth')
    print("Model saved to 'time_series_autoencoder.pth'")
